File: box-and-box/ports/python/score.py
"""score.py - Heuristic Arithmetic, faithful Python port (v0.2)

A Score lives in a SEMIRING (K, +, *, 0, 1): + aggregates alternatives, *
chains evidence. vote/rollout/reinforce/dominate/anneal/softmax satisfy H1-H13.
The semiring zero annihilates * - the algebraic root of the veto used by the
bridge (see bridge.py). Semantics mirror score.mjs exactly.
"""

import math

NEG_INF = float("-inf")


def logsumexp(a, b):
    if a == NEG_INF:
        return b
    if b == NEG_INF:
        return a
    m = max(a, b)
    return m + math.log(math.exp(a - m) + math.exp(b - m))


def _trop_otimes(a, b):
    return NEG_INF if (a == NEG_INF or b == NEG_INF) else a + b


def _log_otimes(a, b):
    return NEG_INF if (a == NEG_INF or b == NEG_INF) else a + b


SEMIRINGS = {
    "tropical": {
        "label": "(max, +)",
        "oplus": lambda a, b: max(a, b),
        "otimes": _trop_otimes,
        "zero": NEG_INF,
        "one": 0,
        "idempotent": True,
    },
    "probability": {
        "label": "(+, x)",
        "oplus": lambda a, b: a + b,
        "otimes": lambda a, b: a * b,
        "zero": 0,
        "one": 1,
        "idempotent": False,
    },
    "log": {
        "label": "(logsumexp, +)",
        "oplus": logsumexp,
        "otimes": _log_otimes,
        "zero": NEG_INF,
        "one": 0,
        "idempotent": False,
    },
}


def Score(p=None):
    """A Score carries a utility plus the soft analogues of the invariant families."""
    p = p or {}
    return {
        "u": p.get("u", 0),          # semiring carrier
        "w": p.get("w", 1),          # [0,1] trust in this heuristic (x, cf. beta's min)
        "eps": p.get("eps", 0),      # [0,1] exploration (anneal -> 0)
        "gamma": p.get("gamma", 1),  # (0,1] discount (soft order, cf. pi's hard order)
        "visits": p.get("visits", 0),  # N under +
        "sources": list(p.get("sources", [])),
    }


def vote(a, b, semiring="tropical"):
    """vote : aggregate alternatives (+ side)."""
    S = SEMIRINGS.get(semiring, SEMIRINGS["tropical"])
    return Score({
        "u": S["oplus"](a["u"], b["u"]),
        "w": a["w"] * b["w"],                # independent trust dilutes (x)
        "eps": max(a["eps"], b["eps"]),
        "gamma": min(a["gamma"], b["gamma"]),
        "visits": a["visits"] + b["visits"],
        "sources": list(a["sources"]) + list(b["sources"]),
    })


def rollout(scores, gamma=0.9, semiring="tropical"):
    """rollout : chain evidence along a path, gamma-discounted (* side).

    score(path) = *_t (gamma^t * u_t). zero anywhere annihilates the whole path.
    """
    S = SEMIRINGS.get(semiring, SEMIRINGS["tropical"])
    acc = S["one"]
    for t, s in enumerate(scores):
        discounted = S["zero"] if s["u"] == S["zero"] else (gamma ** t) * s["u"]
        acc = S["otimes"](acc, discounted)
    return acc


def reinforce(u, target, eta=0.3):
    """reinforce : eta-contraction toward a target."""
    return (1 - eta) * u + eta * target


def dominate(opts):
    """dominate : Pareto-prune (idempotent, antitone).

    opts: [{id, obj:[...]}] higher-is-better; returns the non-dominated front.
    """
    def dominated(a):
        for b in opts:
            if b["id"] != a["id"] \
               and all(bj >= a["obj"][i] for i, bj in enumerate(b["obj"])) \
               and any(bj > a["obj"][i] for i, bj in enumerate(b["obj"])):
                return True
        return False
    return [a for a in opts if not dominated(a)]


def anneal(s):
    """anneal : eps -> 0 (idempotent)."""
    r = Score(s)
    r["eps"] = 0
    return r


def softmax(us, T=1):
    """softmax (shift-invariant; T->0 => argmax)."""
    m = max(us)
    ex = [math.exp((u - m) / T) for u in us]
    z = sum(ex)
    return [e / z for e in ex]

File: box-and-box/ports/go/score.go
package main

// score.go — Heuristic Arithmetic (port of score.mjs).
// A Score lives in a SEMIRING (K, ⊕, ⊗, 0̲, 1̲): ⊕ aggregates alternatives, ⊗ chains
// evidence. 0̲ annihilates ⊗ — the algebraic root of the bridge veto. Laws H1–H13.

import "math"

// Semiring is a commutative-additive / associative-multiplicative dioid.
type Semiring struct {
	Label      string
	Oplus      func(a, b float64) float64
	Otimes     func(a, b float64) float64
	Zero       float64
	One        float64
	Idempotent bool
}

func logsumexp(a, b float64) float64 {
	if math.IsInf(a, -1) {
		return b
	}
	if math.IsInf(b, -1) {
		return a
	}
	m := math.Max(a, b)
	return m + math.Log(math.Exp(a-m)+math.Exp(b-m))
}

func tropicalOtimes(a, b float64) float64 {
	if math.IsInf(a, -1) || math.IsInf(b, -1) {
		return math.Inf(-1)
	}
	return a + b
}

// Semirings registry (tropical / probability / log).
var Semirings = map[string]Semiring{
	"tropical": {
		Label:      "(max, +)",
		Oplus:      func(a, b float64) float64 { return math.Max(a, b) },
		Otimes:     tropicalOtimes,
		Zero:       math.Inf(-1),
		One:        0,
		Idempotent: true,
	},
	"probability": {
		Label:      "(+, ×)",
		Oplus:      func(a, b float64) float64 { return a + b },
		Otimes:     func(a, b float64) float64 { return a * b },
		Zero:       0,
		One:        1,
		Idempotent: false,
	},
	"log": {
		Label:      "(logsumexp, +)",
		Oplus:      logsumexp,
		Otimes:     tropicalOtimes,
		Zero:       math.Inf(-1),
		One:        0,
		Idempotent: false,
	},
}

func semiring(name string) Semiring {
	if s, ok := Semirings[name]; ok {
		return s
	}
	return Semirings["tropical"]
}

// Score carries a utility plus the soft analogues of the invariant families.
type Score struct {
	U       float64 // semiring carrier
	W       float64 // [0,1] trust (×)
	Eps     float64 // [0,1] exploration
	Gamma   float64 // (0,1] discount
	Visits  int     // ℕ under +
	Sources []string
}

// NewScore builds a Score with sensible defaults (matching score.mjs Score()).
func NewScore(u, w, eps, gamma float64, visits int, sources []string) Score {
	if sources == nil {
		sources = []string{}
	}
	return Score{U: u, W: w, Eps: eps, Gamma: gamma, Visits: visits, Sources: append([]string{}, sources...)}
}

// vote : aggregate alternatives (⊕ side).
func vote(a, b Score, semiringName string) Score {
	s := semiring(semiringName)
	return Score{
		U:       s.Oplus(a.U, b.U),
		W:       a.W * b.W,
		Eps:     math.Max(a.Eps, b.Eps),
		Gamma:   math.Min(a.Gamma, b.Gamma),
		Visits:  a.Visits + b.Visits,
		Sources: concatStr(a.Sources, b.Sources),
	}
}

// rollout : chain evidence along a path, γ-discounted (⊗ side).
func rollout(scores []Score, gamma float64, semiringName string) float64 {
	s := semiring(semiringName)
	acc := s.One
	for t, sc := range scores {
		var discounted float64
		if sc.U == s.Zero {
			discounted = s.Zero
		} else {
			discounted = math.Pow(gamma, float64(t)) * sc.U
		}
		acc = s.Otimes(acc, discounted)
	}
	return acc
}

// reinforce : η-contraction toward a target.
func reinforce(u, target, eta float64) float64 { return (1-eta)*u + eta*target }

// Opt is a Pareto candidate (higher-is-better objectives).
type Opt struct {
	ID  int
	Obj []float64
}

// dominate : Pareto-prune (idempotent, antitone).
func dominate(opts []Opt) []Opt {
	out := []Opt{}
	for _, a := range opts {
		dominated := false
		for _, b := range opts {
			if b.ID == a.ID {
				continue
			}
			allGE := true
			someGT := false
			for i := range b.Obj {
				if b.Obj[i] < a.Obj[i] {
					allGE = false
				}
				if b.Obj[i] > a.Obj[i] {
					someGT = true
				}
			}
			if allGE && someGT {
				dominated = true
				break
			}
		}
		if !dominated {
			out = append(out, a)
		}
	}
	return out
}

// anneal : ε → 0 (idempotent).
func anneal(s Score) Score {
	r := s
	r.Eps = 0
	r.Sources = append([]string{}, s.Sources...)
	return r
}

// softmax (shift-invariant; T→0 ⇒ argmax).
func softmax(us []float64, T float64) []float64 {
	m := us[0]
	for _, u := range us {
		if u > m {
			m = u
		}
	}
	ex := make([]float64, len(us))
	z := 0.0
	for i, u := range us {
		ex[i] = math.Exp((u - m) / T)
		z += ex[i]
	}
	out := make([]float64, len(us))
	for i := range ex {
		out[i] = ex[i] / z
	}
	return out
}
